packages/convex-helpers/server/stream.ts

```diff
@@ -0,0 +1,1117 @@
+    middleRange.push([endBoundType, indexFields[0]!, endValue]);
+  }
+  return [...startRanges, middleRange, ...endRanges];
+}
+
+function rangeToQuery(range: Bound[]) {
+  return (q: any) => {
+    for (const [boundType, field, value] of range) {
+      q = q[boundType](field, value);
+    }
+    return q;
+  };
+}
+
+export function getIndexFields<
+  Schema extends SchemaDefinition<any, boolean>,
+  T extends TableNamesInDataModel<DM<Schema>>,
+>(
+  table: T,
+  index?: IndexNames<NamedTableInfo<DM<Schema>, T>>,
+  schema?: Schema,
+): string[] {
+  const indexDescriptor = String(index ?? "by_creation_time");
+  if (indexDescriptor === "by_creation_time") {
+    return ["_creationTime", "_id"];
+  }
+  if (indexDescriptor === "by_id") {
+    return ["_id"];
+  }
+  if (!schema) {
+    throw new Error("schema is required to infer index fields");
+  }
+  const tableInfo = schema.tables[table];
+  const indexInfo = tableInfo.indexes.find(
+    (index: any) => index.indexDescriptor === indexDescriptor,
+  );
+  if (!indexInfo) {
+    throw new Error(`Index ${indexDescriptor} not found in table ${table}`);
+  }
+  const fields = indexInfo.fields.slice();
+  fields.push("_creationTime");
+  fields.push("_id");
+  return fields;
+}
+
+function getIndexKey<
+  DataModel extends GenericDataModel,
+  T extends TableNamesInDataModel<DataModel>,
+>(doc: DocumentByName<DataModel, T>, indexFields: string[]): IndexKey {
+  const key: IndexKey = [];
+  for (const field of indexFields) {
+    let obj: any = doc;
+    for (const subfield of field.split(".")) {
+      obj = obj[subfield];
+    }
+    key.push(obj);
+  }
+  return key;
+}
+
+export function reflect<Schema extends SchemaDefinition<any, boolean>>(
+  db: GenericDatabaseReader<DataModelFromSchemaDefinition<Schema>>,
+  schema: Schema,
+): ReflectDatabaseReader<Schema> {
+  return new ReflectDatabaseReader(db, schema);
+}
+
+/**
+ * A "stream" is an async iterable of query results, ordered by an index on a table.
+ *
+ * Use it as you would use `ctx.db`.
+ * If using pagination in a reactive query, see the warnings on the `paginator`
+ * function. TL;DR: you need to pass in `endCursor` to prevent holes or overlaps
+ * between pages.
+ *
+ * Once you have a stream, you can use `mergeStreams` or `filterStream` to make
+ * more streams. Then use `queryStream` to convert it into an OrderedQuery,
+ * so you can call `.paginate()`, `.collect()`, etc.
+ */
+export function stream<Schema extends SchemaDefinition<any, boolean>>(
+  db: GenericDatabaseReader<DM<Schema>>,
+  schema: Schema,
+): ReflectDatabaseReader<Schema> {
+  return reflect(db, schema);
+}
+
+/**
+ * A "stream" is an async iterable of query results, ordered by an index on a table.
+ */
+export interface IndexStream<
+  DataModel extends GenericDataModel,
+  T extends TableNamesInDataModel<DataModel>,
+> {
+  iterWithKeys(): AsyncIterable<[DocumentByName<DataModel, T> | null, IndexKey]>;
```

Comment:
TODO: IndexKey can contain undefined, make sure we're not serializing it as a convex value